control-plane/agents/core/src/core/specs.rs

```diff
@@ -35,6 +37,416 @@ enum SpecError {
+    where
+        Self: SpecTransaction<O>,
+    {
+        match result {
+            Ok(val) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.commit_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = locked_spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.commit_op();
+                        Ok(val)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(true);
+                        Err(error)
+                    }
+                }
+            }
+            Err(error) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.clear_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = locked_spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.clear_op();
+                        Err(error)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(false);
+                        Err(error)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Start a destroy operation and attempt to log the transaction to the store.
+    /// In case of error, the log is undone and an error is returned.
+    async fn start_destroy<O>(
+        locked_spec: &Arc<Mutex<Self>>,
+        registry: &Registry,
+        del_owned: bool,
+    ) -> Result<(), SvcError>
+    where
+        Self: SpecTransaction<O>,
+        Self: StorableObject,
+    {
+        let mut spec = locked_spec.lock().await;
+        // we're busy with another request, try again later
+        let _ = spec.busy()?;
+        if spec.state().deleted() {
+            Ok(())
+        } else if !del_owned && spec.owned() {
+            Err(SvcError::InUse {
+                kind: spec.kind(),
+                id: spec.uuid(),
+            })
+        } else {
+            spec.set_updating(true);
+            drop(spec);
+
+            // resource specific validation rules
+            if let Err(error) = Self::validate_destroy(&locked_spec, registry).await {
+                let mut spec = locked_spec.lock().await;
+                spec.set_updating(false);
+                return Err(error);
+            }
+            let mut spec = locked_spec.lock().await;
+
+            // once we've started, there's no going back...
+            spec.set_state(SpecState::Deleting);
+
+            spec.start_destroy_op();
+            let spec_clone = spec.clone();
+            drop(spec);
+
+            if let Err(error) = registry.store_obj(&spec_clone).await {
+                let mut spec = locked_spec.lock().await;
+                spec.clear_op();
+                Err(error)
+            } else {
+                Ok(())
+            }
```

Comment:
This pattern repeats a number of times. We could split this out into its own "commit_spec" function.

Reply:
ah yeah it does, I'll give it a go